src/System.Slices/src/System/PtrUtils.cs

```diff
@@ -32,43 +32,50 @@ sealed class PtrUtils
         // depends on it working... (okay, I still feel a little dirty.)
 
         /// <summary>
-        /// Takes a (possibly null) object reference, plus an offset in bytes,
+        /// Takes a (possibly null) object reference, plus an offset in bytes, plus an index,
         /// adds them, and safetly dereferences the target (untyped!) address in
         /// a way that the GC will be okay with.  It yields a value of type T.
         /// </summary>
-
         [MethodImpl(MethodImplOptions.AggressiveInlining)]
-        [ILSub(@"
-            .maxstack 2
-            .locals ([0] uint8& addr)
+        [ILSub(@"            
+            .maxstack 3
+            .locals([0] uint8 & addr)
             ldarg.0     // load the object
             stloc.0     // convert the object pointer to a byref
             ldloc.0     // load the object pointer as a byref
             ldarg.1     // load the offset
             add         // add the offset
+            ldarg.2     // load the index
+            sizeof !!T  // load size of T
+            conv.u  
+            mul         // multiply the index and size of T
+            add         // add the result
             ldobj !!T   // load a T value from the computed address
             ret")]
-        public static T Get<T>(object obj, UIntPtr offset) { return default(T); }
-
+        public static T Get<T>(object obj, UIntPtr offset, UIntPtr index) { return default(T); }
 
         /// <summary>
-        /// Takes a (possibly null) object reference, plus an offset in bytes,
+        /// Takes a (possibly null) object reference, plus an offset in bytes, plus an index,
         /// adds them, and safely stores the value of type T in a way that the
         /// GC will be okay with.
         /// </summary>
-        [MethodImpl(MethodImplOptions.AggressiveInlining)]
-        [ILSub(@"
-            .maxstack 2
-            .locals ([0] uint8& addr)
+        [ILSub(@"            
+            .maxstack 3
+            .locals([0] uint8 & addr)
             ldarg.0     // load the object
             stloc.0     // convert the object pointer to a byref
             ldloc.0     // load the object pointer as a byref
             ldarg.1     // load the offset
             add         // add the offset
-            ldarg.2     // load the value to store
+            ldarg.2     // load the index
+            sizeof !!T  // load size of T
+            conv.u  
+            mul         // multiply the index and size of T
+            add         // add the result
+            ldarg.3     // load the value to store
             stobj !!T   // store a T value to the computed address
             ret")]
-        public static void Set<T>(object obj, UIntPtr offset, T val) { }
+        public static void Set<T>(object obj, UIntPtr offset, UIntPtr index, T val) { }
```

Comment:
Most likely it will. But I'm more curious if we could get rid of the `Set` helper if C# gets ref returns :smile: